service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,372 @@
+                        certificateTagsDao.save(certificateId, tagId);
+                        certTags.add(foundTag);
+                    }
+                } else { //if tag with the same name is not present in db, then save it to db,
+                    // attach to GiftCertificate entity and add to list of passed tags(for return value)
+                    Tag tagEntity = modelMapper.map(tagDto, Tag.class);
+                    Tag savedTag = tagDao.save(tagEntity);
+                    certificateTagsDao.save(certificateId, savedTag.getId());
+                    certTags.add(savedTag);
+                }
+            });
+        }
+
+        List<TagDto> tagsDto = certTags.stream()  //mapping all formed Tag entity to TagDto
+                .map(tag -> modelMapper.map(tag, TagDto.class))
+                .collect(Collectors.toList());
+        GiftCertificateDto dto = modelMapper.map(foundCertificate, GiftCertificateDto.class);//mapping GiftCertificate
+        // entity from db to GiftCertificateDto
+        dto.setTags(tagsDto);//set formed TagDto
+
+        return dto;
+    }
+
+    /**
+     * This method link a tag with a GiftCertificate entity and, if given tag is not present in db, creates it.
+     *
+     * @param tagDto        tag for linking with GiftCertificate entity.
+     * @param certificateId id of GiftCertificate entity.
+     * @return Tag entity from db with id.
+     * @since 1.0
+     */
+    private Tag addTagToCertificate(TagDto tagDto, long certificateId) {
+        Optional<Tag> foundTagOpt = tagDao.getByName(tagDto.getName());
+        Tag tag;
+        if (foundTagOpt.isPresent()) {
+            tag = foundTagOpt.get();
+        } else {
+            Tag tagEntity = modelMapper.map(tagDto, Tag.class);
+            tag = tagDao.save(tagEntity);
+        }
+        Long tagId = tag.getId();
+        certificateTagsDao.save(certificateId, tagId);
+        return tag;
+    }
+
+    /**
+     * This method fills GiftCertificateDto with tags.
+     *
+     * @param giftCertificateDto GiftCertificateDto for filling.
+     * @since 1.0
+     */
+    private void fillCertificateDtoWithTags(GiftCertificateDto giftCertificateDto) {
+        List<Tag> tags = tagDao.getAllByCertificateId(giftCertificateDto.getId());
+        List<TagDto> tagsDto = tags.stream()
+                .map(tag -> modelMapper.map(tag, TagDto.class))
+                .collect(Collectors.toList());
+        giftCertificateDto.setTags(tagsDto);
+    }
+
+    /**
+     * This method throws an exception when GiftCertificate entity with given name and another id is present in db.
+     *
+     * @param name          name of GiftCertificate entity
+     * @param certificateId id of GiftCertificate entity
+     * @throws GiftCertificateWithSuchNameAlreadyExists when there is another GiftCertificate entity in db with given name.
+     */
+    private void exceptionWhenAnotherCertificateWithGivenNameExistsInDb(String name, long certificateId) {
```

Comment:
awful name for the method

Reply:
logic for validator

Reply:
I moved the code of checking an existence or non-existence of an entity to EntityValidator<T> class.
commit 0c4dd35.